src/arcsf/eval/utils.py

```diff
@@ -0,0 +1,299 @@
+        labels: labels (batch_size x sequence_length)
+
+    Returns:
+        _description_
+    """
+    output_logits = output_logits[
+        ..., :-1, :
+    ].contiguous()  # shape: batch_size x (sequence_length-1) x vocab_size
+    shifted_labels = labels[
+        ..., 1:
+    ].contiguous()  # shape : batch_size x (sequence_length - 1)
+    # output_logits.transpose(-1, -2) shape: batch_size x vocab x (sequence_length - 1)
+    loss = _loss_function(output_logits.transpose(-1, -2), shifted_labels).sum(
+        dim=-1
+    )  # loss shape: batch_size
+    target_len = torch.sum(labels != -100, dim=-1)  # length of tokens in target
+    loss_normalised = loss / target_len  # normalised loss shape: batch_size
+    return loss_normalised
+
+
+def get_losses(output, targets):
+    losses = torch.zeros(len(targets))
+    for index, target in enumerate(targets):
+        losses[index] = get_loss(output, target)
+    return losses
+
+
+def ecdf(x):
+    xs, _ = torch.sort(x)
+    ys = torch.arange(1, len(xs) + 1) / float(len(xs))
+    return xs, ys
+
+
+def combine_dicts(forget_dict, retain_dict):
+    return {
+        "retain_tr": retain_dict["truth_ratios"],
+        "forget_tr": forget_dict["truth_ratios"],
+        "rouge_scores": retain_dict["rougeL_recall"],
+    }
+
+
+def get_metrics(
+    base_truth_ratios: torch.Tensor,
+    test_values: dict[torch.Tensor],
+) -> dict[float, float, float, float, float]:
+    """
+    Retrieves metrics for tracking an evaluation.
+
+    Args:
+        base_truth_ratios: base model truth ratios to compare against in the ks_test
+        test_values: recorded values from the evaluation of the test model, including
+        both ks test settings
+
+    Returns:
+        result_dict : dictionary contatining the metrics we would like to track
+    """
+    forget_quality_one_sided = np.log(
+        # the case when the truth ratio CDF for forget is greater than the base model
+        ks_test(test_values["forget_tr"], base_truth_ratios, alternative="greater")
+    )
+    forget_quality_two_sided = np.log(
+        # the general measure of 'closeness' of the truth ratio CDF for forget
+        # against the base model
+        ks_test(test_values["forget_tr"], base_truth_ratios)
+    )
+
+    # need to transform these according to table 1 in the TOFU paper
+    transform_retain_tr = torch.clamp((1 - test_values["retain_tr"]), 0)
+    retain_tr = torch.mean(transform_retain_tr)
+    # calculate other scores
+    rouge_score = torch.mean(test_values["rouge_scores"])
+    model_utilty = hmean([retain_tr, rouge_score])
+
+    result_dict = {
+        "mean_tr_retain": retain_tr.item(),
+        "mean_rouge_score": rouge_score.item(),
+        "forget_quality_1": forget_quality_one_sided.item(),
+        "forget_quality_2": forget_quality_two_sided.item(),
+        "model_utility": model_utilty.item(),
+    }
+
+    return result_dict
+
+
+def all_eval(
+    model: torch.nn.Module,
```

Comment:
All model typehints maybe should be `transformers.PreTrainedModel` (but could be they all work with `torch` modules anyway)

```suggestion
    model: transformers.PreTrainedModel,
```